indico_toolkit/staggered_loop/metrics.py

```diff
@@ -0,0 +1,1431 @@
+"""
+Metrics used for evaluating and comparing partial labels models
+
+Heavily based on:
+* https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/util/metrics.py
+* https://github.com/IndicoDataSolutions/benchner/blob/rd/synthetic/benchner/metrics.py
+
+For all metric calculations below:
+- A "label" or "prediction" is a dictionary of the following format:
+    {"start": int, "end": int, "text": str, "label": str, "doc_idx": int}
+- "true" or "predicted" generally refers to a list of label dictionaries of the format above
+"""
+from collections import defaultdict, OrderedDict
+import copy
+import functools
+import random
+from typing import Callable, Dict, List, Set, Tuple, Union
+
+# Need extra packages to use these metrics
+# TODO Figure out how to incorporate spacy model download into package install
+try:
+    import numpy as np
+    import spacy
+except ImportError:
+    raise Exception("Using partial labels requires installing additional packages")
```

Comment:
Could move those into the setup.py as extras: see https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies and switch to setup.py view in the code block